File: clients/dat-go/v4/dat.go
package dat

import (
	"strconv"
)

type Dat struct {
	data      string
	Expire    uint64
	Cid       uint64
	plainPos  int
	securePos int
	Signature []byte
}

func (d *Dat) Plain() ([]byte, error) {
	return DecodeBase64URL(d.data[d.plainPos : d.securePos-1])
}

func (d *Dat) Secure() ([]byte, error) {
	return DecodeBase64URL(d.data[d.securePos:])
}

func (d *Dat) BodyBytes() []byte {
	return []byte(d.data)
}

func (d *Dat) String() string {
	return strconv.FormatUint(d.Expire, 10) + "." + ToHexFromU64(d.Cid)
}

func ParseDat(s string) (*Dat, error) {
	// single scan: a valid dat has exactly 5 dot-separated parts (4 dots)
	var dots [4]int
	n := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '.' {
			if n == 4 {
				return nil, ErrInvalidDat
			}
			dots[n] = i
			n++
		}
	}
	if n != 4 {
		return nil, ErrInvalidDat
	}

	expire, err := strconv.ParseUint(s[:dots[0]], 10, 64)
	if err != nil || expire <= NowUnixTimestamp() {
		return nil, ErrInvalidDat
	}

	cid, err := strconv.ParseUint(s[dots[0]+1:dots[1]], 16, 64)
	if err != nil {
		return nil, ErrInvalidDat
	}

	signatureB64 := s[dots[3]+1:]
	if signatureB64 == "" {
		return nil, ErrInvalidDat
	}

	signature, err := DecodeBase64URL(signatureB64)
	if err != nil {
		return nil, err
	}

	return &Dat{
		data:      s[:dots[3]],
		Expire:    expire,
		Cid:       cid,
		plainPos:  dots[1] + 1,
		securePos: dots[2] + 1,
		Signature: signature,
	}, nil
}

File: dat-rust/src/dat.rs
use crate::error::DatError;
use crate::util::{decode_base64_url, now_unix_timestamp};
use std::fmt;
use std::str::FromStr;

pub struct Dat {
    dat: Vec<u8>,
    pub(crate) expire: u64,
    pub(crate) cid: u64,
    plain_pos: usize,
    secure_pos: usize,
    pub(crate) signature: Vec<u8>,
}
impl Dat {
    #[inline]
    pub fn expire(&self) -> u64 {
        self.expire
    }
    #[inline]
    pub fn cid(&self) -> u64 {
        self.cid
    }
    #[inline]
    pub(crate) fn plain(&self) -> Result<Vec<u8>, DatError> {
        decode_base64_url(&self.dat[self.plain_pos.. self.secure_pos - 1])
            .map_err(|_| DatError::InvalidDat)
    }
    #[inline]
    pub(crate) fn secure(&self) -> Result<Vec<u8>, DatError> {
        decode_base64_url(&self.dat[self.secure_pos.. ])
            .map_err(|_| DatError::InvalidDat)
    }

    #[inline]
    pub(crate) fn body_bytes(&self) -> &[u8] {
        &self.dat[..]
    }
}

impl fmt::Display for Dat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}.{:x}", self.expire, self.cid)
    }
}

impl FromStr for Dat {
    type Err = DatError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        s.to_string().try_into()
    }
}

impl <'a>TryFrom<&'a str> for Dat {
    type Error = DatError;
    fn try_from(dat: &'a str) -> Result<Self, Self::Error> {
        dat.to_string().try_into()
    }
}

impl TryFrom<String> for Dat {
    type Error = DatError;
    fn try_from(dat: String) -> Result<Self, Self::Error> {
        let mut parts = dat.split('.');

        let expire = parts.next()
            .and_then(|s| s.parse::<u64>().ok())
            .filter(|x| *x > now_unix_timestamp())
            .ok_or_else(|| DatError::InvalidDat)?;

        let cid = parts.next()
            .and_then(|s| u64::from_str_radix(s, 16).ok())
            .ok_or_else(|| DatError::InvalidDat)?;

        let ptr = dat.as_ptr() as usize;
        let plain = parts.next().ok_or_else(|| DatError::InvalidDat)?;
        let plain_pos = plain.as_ptr() as usize - ptr;

        let secure = parts.next().ok_or_else(|| DatError::InvalidDat)?;
        let secure_pos = secure.as_ptr() as usize - ptr;
        let secure_end = secure_pos + secure.len();

        let signature = parts.next().filter(|s| !s.is_empty()).ok_or_else(|| DatError::InvalidDat)?;

        if parts.next().is_some() {
            return Err(DatError::InvalidDat);
        }

        let signature = decode_base64_url(signature)?;
        let mut dat = dat.into_bytes();
        unsafe { dat.set_len(secure_end) };

        Ok(Dat {
            dat,
            expire,
            cid,
            plain_pos,
            secure_pos,
            signature,
        })
    }
}
